### PR TITLE
MINOR: Clean up TestUtils.scala

### PR DESCRIPTION
**Pull request by m1a2st**
Because TestUtils.scala has some unused method, and some methods only use in one class, so I delete unused methods and move methods to individual class which only one class used

### CODE REVIEW

**Comment by m1a2st** [comment]
@chia7712 please review this PR, thank you

**Comment by chia7712** [comment]
@m1a2st Could you please fix the conflicts? thanks!

**Comment by m1a2st** [comment]
@chia7712 I resolve the conflicts

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala [hunk 1]
```diff
@@ -1594,10 +1313,6 @@ object TestUtils extends Logging {
     }
   }
 
-  def createIsrChangeListener(): MockAlterPartitionListener = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/utils/TestUtils.scala:1598]: `MockAlterPartitionListener` is used by `AbstractPartitionTest` only, so could you move it to `AbstractPartitionTest`?

File: core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala [hunk 2]
```diff
@@ -548,4 +549,31 @@ class DeleteTopicTest extends QuorumTestHarness {
...
+  private def increasePartitions[B <: KafkaBroker](admin: Admin,
+                                           topic: String,
+                                           totalPartitionCount: Int,
+                                           brokersToValidate: Seq[B]
+                                          ): Unit = {
+
+    try {
+      val newPartitionSet: Map[String, NewPartitions] = Map.apply(topic -> NewPartitions.increaseTo(totalPartitionCount))
+      admin.createPartitions(newPartitionSet.asJava)
+    } catch {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala]: We do nothing for this catch, so it should be fine to remove it.

File: core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala [hunk 3]
```diff
@@ -548,4 +549,31 @@ class DeleteTopicTest extends QuorumTestHarness {
...
+        throw e
+    }
+
+    if (brokersToValidate.nonEmpty) {
+      // wait until we've propagated all partitions metadata to all brokers
+      val allPartitionsMetadata = waitForAllPartitionsMetadata(brokersToValidate, topic, totalPartitionCount)
+
+      (0 until totalPartitionCount - 1).map { i =>
+        i -> allPartitionsMetadata.get(new TopicPartition(topic, i)).map(_.leader()).getOrElse(
+          throw new IllegalStateException(s"Cannot get the partition leader for topic: $topic, partition: $i in server metadata cache"))
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala]: this method return nothing, so we don't need to collect the metadata. However, it needs to check the existence of topic partition

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala [hunk 4]
```diff
@@ -613,4 +613,19 @@ class LogSegmentTest {
...
+    assertFalse(s1.hasNext, "Iterators have uneven length--first has more")
+    assertFalse(s2.hasNext, "Iterators have uneven length--second has more")
+  }
+
+  private def writeNonsenseToFile(fileName: File, position: Long, size: Int): Unit = {
+    val file = new RandomAccessFile(fileName, "rw")
+    file.seek(position)
+    for (_ <- 0 until size)
+      file.writeByte(random.nextInt(255))
+    file.close()
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: could you please add try-finally?

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala [hunk 5]
```diff
@@ -52,6 +56,20 @@ class SaslPlainPlaintextConsumerTest extends BaseConsumerTest with SaslSetup {
...
-    TestUtils.verifyUnsecureZkAcls(zkClient)
+    verifyUnsecureZkAcls(zkClient)
+  }
+
+  /**
+   * Verifies that secure paths in ZK have no access control. This is
+   * the case when zookeeper.set.acl=false and no ACLs have been configured.
+   */
+  private def verifyUnsecureZkAcls(zkClient: KafkaZkClient): Unit = {
+    secureZkPaths(zkClient).foreach(path => {
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:58]: We can merge this method into test case `testZkAclsDisabled`

File: core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala [hunk 6]
```diff
@@ -149,6 +150,23 @@ class SaslPlainSslEndToEndAuthorizationTest extends SaslEndToEndAuthorizationTes
...
   @Test
   def testAcls(): Unit = {
-    TestUtils.verifySecureZkAcls(zkClient, 1)
+    verifySecureZkAcls(zkClient, 1)
+  }
+
+  /**
+   * Verifies that all secure paths in ZK are created with the expected ACL.
+   */
+  private def verifySecureZkAcls(zkClient: KafkaZkClient, usersWithAccess: Int): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala]: ditto

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for nice cleanup! a couple of comments left. PTAL

**Comment by m1a2st** [comment]
@chia7712 thanks for your comment, I have already change these code by your comment

**Comment by chia7712** [comment]
@m1a2st Could you please fix conflicts?

**Comment by chia7712** [comment]
```
[2024-04-27T01:29:26.736Z] > Task :core:compileTestScala

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:20:33: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:20:40: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:25:25: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:31:17: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:34:40: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:34:56: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:53:49: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:58:46: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:58:62: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:58:86: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:58:109: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:59:33: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:60:30: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:63:40: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:63:60: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:66:43: Unused import

[2024-04-27T01:29:26.736Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:68:54: Unused import

[2024-04-27T01:29:28.026Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/utils/TestUtils.scala:61:33: Unused import

[2024-04-27T01:29:28.026Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/utils/TestUtils.scala:61:55: Unused import

[2024-04-27T01:29:28.026Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/utils/TestUtils.scala:63:54: Unused import

[2024-04-27T01:29:28.026Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/utils/TestUtils.scala:63:74: Unused import

[2024-04-27T01:29:28.026Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/utils/TestUtils.scala:63:92: Unused import

[2024-04-27T01:29:28.026Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15808/core/src/test/scala/unit/kafka/utils/TestUtils.scala:63:106: Unused import

[2024-04-27T01:29:29.239Z] 23 errors found

```

please fix the build error :)

**Comment by m1a2st** [comment]
I have been remove these unused imports, please review thanks

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala [hunk 7]
```diff
@@ -1124,24 +933,18 @@ object TestUtils extends Logging {
...
-   * @return The unwrapped value returned by the function
-   */
-  def awaitValue[T](func: () => Option[T], msg: => String, waitTimeMs: Long = JTestUtils.DEFAULT_MAX_WAIT_MS): T = {
-    var value: Option[T] = None
-    waitUntilTrue(() => {
-      value = func()
-      value.isDefined
-    }, msg, waitTimeMs)
-    value.get
+  def subscribeAndWaitForRecords(topic: String,
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/utils/TestUtils.scala]: why adding this new method? it is unused.

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for updated PR. The following methods seems to be moved to callers also. Please help us do double-check.

1. `setReplicationThrottleForPartitions`
2. `removeReplicationThrottleForPartitions`
3. `removeReplicationThrottleForPartitions`
4. `stringifyTopicPartitions`
5. `assertBadConfigContainingMessage`
6. `makeLeaderForPartition`
7. `getMsgStrings`

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala (hunk 7 above)
Comment (m1a2st) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: `assertBadConfigContainingMessage` this method are used by two classes, so I didn't move it. 
The `kafka.KafkaConfigTest` and `server.KafkaConfigTest` are using it

Comment (m1a2st) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: Other methods refactored by your comment.

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala [hunk 8]
```diff
@@ -1086,4 +1087,10 @@ class RemoteIndexCacheTest {
       case e @ (_ : NoSuchFileException | _ : UncheckedIOException) => Optional.empty()
     }
   }
+
+  private def numThreadsRunning(threadNamePrefix: String, isDaemon: Boolean): mutable.Set[Thread] = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: We can fix the build error by rewriting this function via java lambda. For example:
```scala
  private def numThreadsRunning(threadNamePrefix: String, isDaemon: Boolean): java.util.Set[Thread] = {
    Thread.getAllStackTraces.keySet.stream().filter { t =>
      isDaemon && t.isAlive && t.getName.startsWith(threadNamePrefix)
    }.collect(Collectors.toSet)
  }
```

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala [hunk 9]
```diff
@@ -23,25 +23,26 @@ import org.apache.kafka.common.{TopicIdPartition, TopicPartition, Uuid}
...
 import org.junit.jupiter.params.provider.EnumSource
 import org.mockito.ArgumentMatchers
 import org.mockito.ArgumentMatchers.any
-import org.mockito.invocation.InvocationOnMock
 import org.mockito.Mockito._
+import org.mockito.invocation.InvocationOnMock
 import org.slf4j.{Logger, LoggerFactory}
 
-import java.io.{File, FileInputStream, IOException, PrintWriter, UncheckedIOException}
+import java.io._
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: We should avoid using `_` since it could cause naming conflicts.

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for updated PR. two major comments left. Please take a look.

**Comment by chia7712** [comment]
@m1a2st please make sure the build gets work with scala 2.12. You can use following command: `./gradlew clean build -x test -PscalaVersion=2.12`

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala (hunk 8 above)
Comment (m1a2st) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: I change the function name, and modify using java set
```java  
private def getRunningCleanerThread: java.util.Set[Thread] = {
    Thread.getAllStackTraces.keySet()
      .stream()
      .filter(t => t.isAlive && t.getName.startsWith(REMOTE_LOG_INDEX_CACHE_CLEANER_THREAD))
      .collect(Collectors.toSet())
  }
  ```

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala [hunk 10]
```diff
@@ -321,16 +322,16 @@ class RemoteIndexCacheTest {
     // Give the thread cleaner thread some time to throw an exception
     Thread.sleep(100)
     // Verify that Cleaner thread is still running even when exception is thrown in doWork()
-    var threads = TestUtils.numThreadsRunning(REMOTE_LOG_INDEX_CACHE_CLEANER_THREAD, isDaemon = true)
+    var threads = getRunningCleanerThread
     assertEquals(1, threads.size,
-      s"Found unexpected ${threads.size} threads=${threads.map(t => t.getName).mkString(", ")}")
+      s"Found unexpected ${threads.size} threads=${threads.stream().map(t => t.getName).collect(Collectors.joining(", "))}")
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: `threads.stream().map[String](t => t.getName).collect(Collectors.joining(", "))`

You can define the `type` explicitly to fix the type mismatch in scala 2.12

Comment (m1a2st) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: fix it, and run `./gradlew clean build -x test -PscalaVersion=2.12` is build successful

Comment (chia7712) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: nice job!

**Review by chia7712 — APPROVED** [review]
LGTM

```
./gradlew cleanTest :streams:test --tests EOSUncleanShutdownIntegrationTest.shouldWorkWithUncleanShutdownWipeOutStateStore :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful --tests GetOffsetShellTest.testTopicPartitionsArgWithInternalIncluded :storage:test --tests TransactionsWithTieredStoreTest.testSendOffsetsWithGroupMetadata --tests TransactionsWithTieredStoreTest.testFailureToFenceEpoch --tests TransactionsWithTieredStoreTest.testSendOffsetsToTransactionTimeout :connect:runtime:test --tests org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testSeparateOffsetsTopic :metadata:test --tests QuorumControllerTest.testFenceMultipleBrokers :streams:examples:test --tests WordCountDemoTest.testOneWord --tests WordCountDemoTest.testCountListOfWords --tests WordCountDemoTest.testGetStreamsConfig :connect:mirror:test --tests IdentityReplicationIntegrationTest.testSyncTopicConfigs --tests IdentityReplicationIntegrationTest.testOffsetSyncsTopicsOnTarget --tests MirrorConnectorsIntegrationBaseTest.testReplicationWithEmptyPartition --tests MirrorConnectorsIntegrationBaseTest.testSyncTopicConfigs --tests MirrorConnectorsIntegrationTransactionsTest.testSyncTopicConfigs :core:test --tests AddPartitionsTest.testIncrementPartitions --tests ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit --tests TransactionsTest.testFencingOnCommit :clients:test --tests KafkaConsumerTest.testShouldAttemptToRejoinGroupAfterSyncGroupFailed
```
loop them 3 times, and all pass. will merge this PR

**Pull request merged**